zipkin-server/src/main/java/zipkin2/server/internal/throttle/ThrottledCall.java

```diff
@@ -0,0 +1,211 @@
+ * limitations under the License.
+ */
+package zipkin2.server.internal.throttle;
+
+import com.netflix.concurrency.limits.Limiter;
+import com.netflix.concurrency.limits.Limiter.Listener;
+import java.io.IOException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.function.Supplier;
+import zipkin2.Call;
+import zipkin2.Callback;
+import zipkin2.storage.InMemoryStorage;
+
+/**
+ * {@link Call} implementation that is backed by an {@link ExecutorService}. The ExecutorService serves two
+ * purposes:
+ * <ol>
+ * <li>Limits the number of requests that can run in parallel.</li>
+ * <li>Depending on configuration, can queue up requests to make sure we don't aggressively drop requests that would
+ *     otherwise succeed if given a moment. Bounded queues are safest for this as unbounded ones can lead to heap
+ *     exhaustion and {@link OutOfMemoryError OOM errors}.</li>
+ * </ol>
+ *
+ * @see ThrottledStorageComponent
+ */
+class ThrottledCall<V> extends Call<V> {
+  private final ExecutorService executor;
+  private final Limiter<Void> limiter;
+  private final Listener limitListener;
+  /**
+   * Delegate call needs to be supplied later to avoid having it take action when it is created (like
+   * {@link InMemoryStorage} and thus avoid being throttled.
+   */
+  private final Supplier<Call<V>> delegate;
+  private Call<V> call;
+  private boolean canceled;
+
+  public ThrottledCall(ExecutorService executor, Limiter<Void> limiter, Supplier<Call<V>> delegate) {
+    this.executor = executor;
+    this.limiter = limiter;
+    this.limitListener = limiter.acquire(null).orElseThrow(RejectedExecutionException::new);
+    this.delegate = delegate;
+  }
+
+  private ThrottledCall(ThrottledCall other) {
+    this(other.executor, other.limiter, other.call == null ? other.delegate : () -> other.call.clone());
+  }
+
+  @Override
+  public V execute() throws IOException {
+    try {
+      call = delegate.get();
+
+      // Make sure we throttle
+      Future<V> future = executor.submit(() -> {
+        try (AutoCloseable nameReverter = updateThreadName(call.toString())) {
+          return call.execute();
+        }
+      });
+      V result = future.get(); // Still block for the response
+
+      limitListener.onSuccess();
+      return result;
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof RejectedExecutionException) {
+        // Storage rejected us, throttle back
+        limitListener.onDropped();
+      } else {
+        limitListener.onIgnore();
+      }
+
+      throw new ThrottleException(cause);
+    } catch (RuntimeException e) {
+      limitListener.onIgnore();
+      throw e; // E.g. RejectedExecutionException
+    } catch (Exception e) {
```

Comment:
IOException is never thrown in the corresponding `try` block.  It would wrapped up in the ExecutionException so I updated the usage of ThrottleException on line 91 to account for this.

Reply:
roger. I guess the fact this is catching Exception is what threw me... did you want to catch `(RuntimeException|Error)` here or is something actually declaring exception throws.